### ISSUE TITLE
Platform.Select not working for ios builds

### ISSUE THREAD

**Issue by jinxac** [facebook/react-native#17566]
### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

  Environment:
    OS:  macOS Sierra 10.12.6
    Node: 7.2.1
    Yarn:  0.27.5
    npm:  3.10.10
    Watchman:  4.7.0
    Xcode:  Xcode 8.3.3 Build version 8E3004b

  Packages: (wanted => installed)
    react-native: 0.48.1 => 0.47.2
    react: 16.0.0-alpha.12 => 16.0.0-alpha.12

### Steps to Reproduce

I am observing something strange with `Platform.Select` behavior on iOS.  

I recently launched referral feature and have following piece of code to save referrer.

```
  static saveReferrer = (referrer) => {
    return Platform.select({
      ios: AsyncStorage.setItem(IOS_REFERRER_STORAGE_KEY, referrer),
      android: Promise.reject(new Error("Feature unavailable on android"))
    });
  };
```

As seen for ios I want to set item  in my AsyncStore but in android I handle it a different way so i just skip it. Now I am seeing that while using above code android platform code ( `  android: Promise.reject(new Error("Feature unavailable on android")) `) is getting returned for iOS.  On android this works fine howerver.

I tried changing the code above to following and it works fine. 

```
  static saveReferrer = (referrer) => {
    if (Platform.OS === "ios") {
       AsyncStorage.setItem(IOS_REFERRER_STORAGE_KEY, referrer);
    } else {
       Promise.reject(new Error("Feature unavailable on android"));
     }
  };
```

![screen shot 2018-01-12 at 4 14 54 pm](https://user-images.githubusercontent.com/15068096/34871814-f0b7adc8-f7b4-11e7-8799-857955cf8488.png)

### Expected Behavior

Platform.Select should  call platform specific code in case of iOS

Attaching screenshot for the same issue

**Comment by Bhullnatik** [comment]
You are passing the result of functions to `Platform.select()`, not functions. So in order to compute the object, JavaScript is calling both of the methods _and is then_ setting the result to `ios` and `android`. Basically, all your code is executed on every platform.

You could do something like:
```javascript
return Platform.select({
  ios: () => AsyncStorage.setItem(IOS_REFERRER_STORAGE_KEY, referrer),
  android: () => Promise.reject(new Error("Feature unavailable on android"))
})();
```

But that's kind of hard to follow imho, Your second approach is better.

**Comment by jinxac** [comment]
yeah @Bhullnatik you are right
